### PR TITLE
fix: JSON schema `examples` were OpenAPI formatted

### PR DESCRIPTION
**Pull request by tuukkamustonen**
## Description

The generated `examples` in *JSON schema* objects were formatted as:

```json
    "examples": {
      "some-id": {
        "description": "Lorem ipsum",
        "value": "the real beef"
      }
   }
```

However, above is OpenAPI example format, and must not be used in JSON schema
objects. Schema objects follow different formatting:

```json
    "examples": [
      "the real beef"
   ]
```

This is explained in
https://medium.com/apis-you-wont-hate/openapi-v3-1-and-json-schema-2019-09-6862cf3db959

Schema objects spec in
https://spec.openapis.org/oas/v3.1.0#schema-object.

OpenAPI example format spec in
https://spec.openapis.org/oas/v3.1.0#example-object. This is referenced at
least from parameters, media types and components.

The technical change here is to define `Schema.examples` as `list[Any]` instead
of `list[Example]`. Examples can and must still be defined as `list[Example]`
for OpenAPI objects (e.g. `Parameter`, `Body`) but for JSON schema `examples`
the code now internally generates/converts `list[Any]` format instead.

Extra confusion here comes from the OpenAPI 3.0 vs OpenAPI 3.1 difference.
OpenAPI 3.0 only allowed `example` (singular) field in schema objects.
OpenAPI 3.1 supports the full JSON schema 2020-12 spec and so `examples` array
in schema objects.

Both `example` and `examples` seem to be supported, though the former is marked
as deprecated in the latest specs.

This can be tested over at https://editor-next.swagger.io by loading up the
OpenAPI 3.1 Pet store example. Then add `examples` in `components.schemas.Pet`
using the both ways and see the Swagger UI only render the example once it's 
properly formatted (it ignores is otherwise).

## Closes

n/a

Partially fixes https://github.com/litestar-org/litestar/issues/2849 (the OpenAPI UIs don't render badly formatted examples in the spec).

### CODE REVIEW

**Comment by guacs** [comment]
@tuukkamustonen thanks for this! This is something I'd noticed and was meaning to fix, but hadn't gotten the time to actually investigate.

**Comment by tuukkamustonen** [comment]
Let's see if I can be type linters pass... interesting errors...

**Comment by provinzkraut** [comment]
Is this a regression @guacs? I vaguely remember a PR that fixed a similar (or actually the same) bug?

**Comment by guacs** [comment]
> Is this a regression @guacs? I vaguely remember a PR that fixed a similar (or actually the same) bug?

If I remember correctly, that PR was for the rendering of examples of things like path parameters, query parameters etc. This is for the examples of the schemas we generate.

**Comment by tuukkamustonen** [comment]
I'm getting these errors even on latest `main`:

```
=> Running mypy
tests/e2e/test_pydantic.py:11: error: Function is missing a type annotation  [no-untyped-def]
tests/unit/test_contrib/test_pydantic/test_schema_plugin.py:73: error: Function is missing a type annotation  [no-untyped-def]
tests/unit/test_openapi/test_responses.py:1: error: Cannot find implementation or library stub for module named "litestar.openapi.plugins"  [import-not-found]
tests/unit/test_openapi/test_responses.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
tests/unit/test_security/test_security.py:1: error: Cannot find implementation or library stub for module named "litestar.openapi.plugins"  [import-not-found]
Found 4 errors in 4 files (checked 691 source files)
```

In `develop`:

```
=> Running mypy
tests/e2e/test_pydantic.py:11: error: Function is missing a type annotation  [no-untyped-def]
tests/unit/test_contrib/test_pydantic/test_schema_plugin.py:73: error: Function is missing a type annotation  [no-untyped-def]
Found 2 errors in 2 files (checked 697 source files)
```

**Comment by tuukkamustonen** [comment]
Worked after rebase on latest `main` now, okay...

**Comment by peterschutt** [comment]
> Is this a regression @guacs? I vaguely remember a PR that fixed a similar (or actually the same) bug?

I think this #2660 which was a fix for #2272

File: litestar/typing.py [hunk 1]
```diff
@@ -81,10 +92,11 @@ def _parse_metadata(value: Any, is_sequence_container: bool, extra: dict[str, An
         **cast("dict[str, Any]", extra or getattr(value, "extra", None) or {}),
         **(getattr(value, "json_schema_extra", None) or {}),
     }
+    example_list: list[Any] | None
     if example := extra.pop("example", None):
-        example_list = [Example(value=example)]
+        example_list = [example]
     elif examples := getattr(value, "examples", None):
-        example_list = [Example(value=example) for example in cast("list[str]", examples)]
+        example_list = examples
```
[the remark this answers is not in the document]
Comment (peterschutt) [reply litestar/typing.py:99]: Looking at where _we_ use it, changing the type of `KwargDefinition.examples` likely isn't that big a deal, but it is publicly documented as it is, and I don't know that changing its type to `Any` is very good.

I think this also means that `get_formatted_examples()` could receive `Sequence[Any]` now, is that right?

https://github.com/litestar-org/litestar/blob/024dccff1e6bf433d260c7fd2add7694eb05f63f/litestar/_openapi/schema_generation/utils.py#L103

File: litestar/_openapi/schema_generation/schema.py [hunk 2]
```diff
@@ -558,7 +558,7 @@ def process_schema_result(self, field: FieldDefinition, schema: Schema) -> Schem
                     not isinstance(value, Hashable) or not self.is_undefined(value)
                 ):
                     if schema_key == "examples":
-                        value = get_formatted_examples(field, cast("list[Example]", value))
+                        value = get_json_schema_examples(field, cast("list[Example]", value))
```
Comment (peterschutt) [inline litestar/_openapi/schema_generation/schema.py:561]: The `cast` here would be `list[Any]` assuming we changed the type of `KwargDefinition.examples`

File: litestar/_openapi/schema_generation/utils.py [hunk 3]
```diff
@@ -107,3 +107,14 @@ def get_formatted_examples(field_definition: FieldDefinition, examples: Sequence
     name = name.lower()
 
     return {f"{name}-example-{i}": example for i, example in enumerate(examples, 1)}
+
+
+def get_json_schema_examples(field_definition: FieldDefinition, examples: Sequence[Example | str]) -> list[Any]:
```
Comment (peterschutt) [inline litestar/_openapi/schema_generation/utils.py:112]:
```suggestion
def get_json_schema_examples(examples: Sequence[Example | str]) -> list[Any]:
```
Wonder why linters didn't get this.

**Comment by peterschutt** [comment]
This was the MCVE from #2272:

```py
from typing_extensions import Annotated
from msgspec import Struct, Meta
from litestar import Litestar, post


class TestBody(Struct):
    field1: Annotated[str, Meta(examples=["the answer"])]
    field2: Annotated[int, Meta(examples=[42])]


@post(sync_to_thread=False)
def test(data: TestBody) -> None:
    ...


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app=Litestar(route_handlers=[test]), port=8080)
```

If I run that app on `main` we get:
image

Which doesn't include the example values for those fields, so either we've had a regression or that wasn't fixed.

Same app on this branch:

image

So LGTM on that front.

**Comment by guacs** [comment]
> This was the MCVE from #2272:
> 
> ```python
> from typing_extensions import Annotated
> from msgspec import Struct, Meta
> from litestar import Litestar, post
> 
> 
> class TestBody(Struct):
>     field1: Annotated[str, Meta(examples=["the answer"])]
>     field2: Annotated[int, Meta(examples=[42])]
> 
> 
> @post(sync_to_thread=False)
> def test(data: TestBody) -> None:
>     ...
> 
> 
> if __name__ == "__main__":
>     import uvicorn
>     uvicorn.run(app=Litestar(route_handlers=[test]), port=8080)
> ```
> 
> If I run that app on `main` we get: image
> 
> Which doesn't include the example values for those fields, so either we've had a regression or that wasn't fixed.
> 
> Same app on this branch:
> 
> image
> 
> So LGTM on that front.

@peterschutt how were you able to verify that things are working with the fix introduced in the linked PR? I tried running the app in this comment after checking out to `680c2df00`, but I still was getting this incorrect example. I think I may be doing something wrong.

File: litestar/_openapi/schema_generation/schema.py (hunk 2 above)
Comment (tuukkamustonen) [reply litestar/_openapi/schema_generation/schema.py:561]: The `cast` here is for the `KwargDefinition.examples` field, which still is declared as `list[Example] | None` and passed into `get_json_schema_examples()`. It _should_ be correct afaik.

The return value is `list[Any]` which is for JSON schema (and not OpenAPI) examples. Also that _should_ be correct.

File: litestar/_openapi/schema_generation/utils.py (hunk 3 above)
Comment (tuukkamustonen) [reply litestar/_openapi/schema_generation/utils.py:112]: Ha, indeed...

File: litestar/typing.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tuukkamustonen) [reply litestar/typing.py:99]: The method here picks up stuff from the field declaration, and puts into that dict what is relevant / needed to create... at least `Schema` entry <-- I think my understanding/assumption was that it's only used for `Schema` creation, but if it's also used for something else(?) then there may be something wrong here.

But that's different than the flow that you described.

I'm not changing `KwargDefinition.examples` type and must not change it. It will remain as `list[Example]`, as that's the syntax/format how _OpenAPI_ examples are declared currently and that's fine. But JSON schema `examples` need that `list[Any]` instead (as they're just arrays of values).

I guess the question here is exactly what this `_parse_metadata()` is used for - where is that return value used... I'll check again.

> I find this corner of the library really hard to follow

Yes, I would even describe it as _super_ complex and kludgy 😵‍💫

[the remark this answers is not in the document]
Comment (tuukkamustonen) [reply litestar/typing.py]: Yeah, I believe you're righting. I reverted that in a fixup commit.

Funnily, changing it one way or around doesn't break anything. For sure there's _some_ impact (in the OpenAPI output), but I guess it's not unit test covered.

**Comment by tuukkamustonen** [comment]
@peterschutt Fixed as suggested in the fixup commit. Great that you spotted that! Take another look.

@guacs I also tried the MVCE there, and it does work for me 🤷🏻‍♂️

**Comment by peterschutt** [comment]
> @peterschutt how were you able to verify that things were working with the fix introduced in the linked PR? I tried running the app in this comment after checking out to `680c2df00`, but I still was getting this incorrect example. I think I may be doing something wrong.

My verification was that it was incorrect on `main` and `correct` on this branch - given that you get the incorrect behavior on `680c2df00`, and I get same on `main` and `680c2df00` was merged into main, I think we are saying the same thing. Unless I'm misunderstanding (very possible :D)

File: litestar/typing.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (peterschutt) [reply litestar/typing.py]: > but if it's also used for something else(?)

Its used on any occasion that we parse an annotation that has _any object_ that defines those openapi/schema constraint attributes in its metadata, e.g., `query_value: Annotated[int, Parameter(example=...)]` on a handler, and also `field_on_a_model: Annotated[int, Meta(example=...)]` on some domain model. Both of these cases get end up with the example value getting shoehorned into `Example`, which - as your recent work has brought to light - is not a correct abstraction for the case where the example should exist on a json schema document.

However imperfect it is that we are using `Example` to carry around that json schema example value internally here between model parsing and schema generation - I guess "it is what it is" for now.

> Yes, I would even describe it as super complex and kludgy 😵‍💫

`FieldDefinition` used to be called `ParsedType` - and its only purpose was to represent a type annotation. It would handle extraction of meta values out of `Annotated`, but not make any assumptions about how that metadata should be used - just a container for type information.

I don't think we need to do all of this metadata parsing/merging/transforming here because its got nothing to do with parsing type annotations and everything to do with schema generation, so should probably exist in `litestar._openapi`. Problem there is that at the same time `ParsedType` turned into `FieldDefinition`, `FieldDefinition.kwarg_definition` was added to the public api, which is the result of all of this metadata wrangling. So its probably going to have to be a 3.0 thing to unwind.

File: litestar/_openapi/schema_generation/schema.py (hunk 2 above)
Comment (peterschutt) [reply litestar/_openapi/schema_generation/schema.py]: My understanding is that it is correct now given the rollback on changes to metadata parsing - my point was that we were parsing the data that feeds `KwargDefinition.examples` with `list[Any]`.. anyhow I think its moot now.

**Review by peterschutt — APPROVED** [review]
Thanks @tuukkamustonen!

**Comment by winstxnhdw** [comment]
Hey, is there a reason why this doesn't work when using class-based controllers?
